protocols/randhound/randhound.go

```diff
@@ -1,251 +1,995 @@
-	Time        time.Time // Scheduled initiation time
+	// Setup message handlers
+	h := []interface{}{
+		rh.handleI1, rh.handleI2,
+		rh.handleR1, rh.handleR2,
+	}
+	err := rh.RegisterHandlers(h...)
+
+	return rh, err
 }
 
-// Group encapsulates all the configuration parameters of a list of RandHound nodes.
-type Group struct {
-	N uint32 // Total number of nodes (peers + leader)
-	F uint32 // Maximum number of Byzantine nodes tolerated (1/3)
-	L uint32 // Minimum number of non-Byzantine nodes required (2/3)
-	K uint32 // Total number of trustees (= shares generated per peer)
-	R uint32 // Minimum number of signatures needed to certify a deal
-	T uint32 // Minimum number of shares needed to reconstruct a secret
+// Setup configures a RandHound instance on client-side. Needs to be called
+// before Start.
+func (rh *RandHound) Setup(nodes int, faulty int, groups int, purpose string) error {
+
+	rh.nodes = nodes
+	rh.groups = groups
+	rh.faulty = faulty
+	rh.purpose = purpose
+
+	rh.server = make([][]*sda.TreeNode, groups)
+	rh.group = make([][]int, groups)
+	rh.threshold = make([]int, groups)
+	rh.key = make([][]abstract.Point, groups)
+	rh.ServerIdxToGroupNum = make([]int, nodes)
+	rh.ServerIdxToGroupIdx = make([]int, nodes)
+
+	rh.i1s = make(map[int]*I1)
+	rh.i2s = make(map[int]*I2)
+	rh.r1s = make(map[int]*R1)
+	rh.r2s = make(map[int]*R2)
+	rh.polyCommit = make(map[int][]abstract.Point)
+	rh.secret = make(map[int][]int)
+	rh.chosenSecret = make(map[int][]int)
+
+	rh.Done = make(chan bool, 1)
+	rh.SecretReady = false
+
+	return nil
 }
 
-// Leader (=client) refers to the node which initiates the RandHound protocol,
-// moves it forward, and ultimately outputs the generated public randomness.
-type Leader struct {
-	rc      []byte                 // Leader's trustee-selection random value
-	rs      [][]byte               // Peers' trustee-selection random values
-	i1      *I1                    // I1 message sent to the peers
-	i2      *I2                    // I2 - " -
-	i3      *I3                    // I3 - " -
-	i4      *I4                    // I4 - " -
-	r1      map[uint32]*R1         // R1 messages received from the peers
-	r2      map[uint32]*R2         // R2 - " -
-	r3      map[uint32]*R3         // R3 - " -
-	r4      map[uint32]*R4         // R4 - " -
-	states  map[uint32]*poly.State // States for deals and responses from peers
-	invalid map[uint32]*[]uint32   // Map to mark invalid shares
-	Done    chan bool              // For signaling that a protocol run is finished
+// Start initiates the RandHound protocol run. The client pseudo-randomly
+// chooses the server grouping, forms an I1 message for each group, and sends
+// it to all servers of that group.
+func (rh *RandHound) Start() error {
+
+	var err error
+
+	// Set timestamp
+	rh.time = time.Now()
+
+	// Choose client randomness
+	hs := rh.Suite().Hash().Size()
+	rand := make([]byte, hs)
+	random.Stream.XORKeyStream(rand, rand)
+	rh.cliRand = rand
```

Comment:
Ah thanks I didn't know about this function in dedis/crypto, I will of course use that instead.